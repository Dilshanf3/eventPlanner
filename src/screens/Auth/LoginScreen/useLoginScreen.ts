import {useState} from 'react';
import {useNavigation} from '@react-navigation/native';
import {isValidEmail} from '../../Utils/validationUtils';
import {
  signUpWithEmailAndPassword,
  signInWithEmailAndPassword,
} from '../../../services/LoginService/authService';
import ScreenNames from '../../../constants/screenNames';
import {RootStackParamList} from '../../../types';
import {NativeStackNavigationProp} from 'react-native-screens/lib/typescript/native-stack/types';

type NavigationProp = NativeStackNavigationProp<RootStackParamList>;

const useLoginScreen = () => {
  const navigation = useNavigation<NavigationProp>();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const [showPassword, setShowPassword] = useState(false);
  const [showConfirmPassword, setShowConfirmPassword] = useState(false);
  const [isSignUp, setIsSignUp] = useState(false);
  const [loading, setLoading] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');
 

  const toggleShowPassword = () => setShowPassword(!showPassword);
  const toggleShowConfirmPassword = () =>
    setShowConfirmPassword(!showConfirmPassword);
  const handleSignUpToggle = () => setIsSignUp(!isSignUp);
  const handleLoginToggle = () => {
    setIsSignUp(false);
    setErrorMessage('');
  };

  const resetForm = () => {
    setEmail('');
    setPassword('');
    setConfirmPassword('');
    setErrorMessage('');
  };

  const handleSignUp = async () => {
    if (
      email.trim() === '' ||
      password.trim() === '' ||
      confirmPassword.trim() === ''
    ) {
      setErrorMessage('Required fields cannot be empty');
      return;
    }

    if (!isValidEmail(email)) {
      setErrorMessage('Please enter a valid email address.');
      return;
    }

    if (password !== confirmPassword) {
      setErrorMessage('Passwords do not match!');
      return;
    }

    setLoading(true);

    try {
      await signUpWithEmailAndPassword(email, password); // Use the utility function for sign-up
      resetForm(); // Reset form fields after successful sign-up
      navigation.navigate(ScreenNames.PROFILE_IMAGE);
    } catch (error) {
      setErrorMessage(error.message);
    } finally {
      setLoading(false);
    }
  };

  const handleLogin = async () => {
    if (email.trim() === '' || password.trim() === '') {
      setErrorMessage('Please fill in all fields.');
      return;
    }

    if (!isValidEmail(email)) {
      setErrorMessage('Please enter a valid email address.');
      return;
    }

    setLoading(true);

    try {
      await signInWithEmailAndPassword(email, password); // Use the utility function for login
      resetForm(); // Reset form fields after successful login
      navigation.navigate(ScreenNames.DASHBOARD);
    } catch (error) {
      setErrorMessage(error.message);
    } finally {
      setLoading(false);
    }
  };

  return {
    email,
    password,
    confirmPassword,
    showPassword,
    showConfirmPassword,
    isSignUp,
    loading,
    errorMessage,
    setEmail,
    setPassword,
    setConfirmPassword,
    toggleShowPassword,
    toggleShowConfirmPassword,
    handleSignUpToggle,
    handleLoginToggle,
    handleSignUp,
    handleLogin,
  };
};

export default useLoginScreen;
