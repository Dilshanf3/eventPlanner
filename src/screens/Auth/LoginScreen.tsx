import React from 'react';
import {Text, TouchableOpacity, Image} from 'react-native';
import {SafeAreaView} from 'react-native-safe-area-context';
import InputField from '../../component/InputField';
import useLoginScreen from './useLoginScreen';
import Spinner from 'react-native-loading-spinner-overlay';
import styles from '../../styles/loginStyles';
import {Strings} from '../../constants/strings';
import ButtonComponent from '../../component/Button/ButtonComponent';
import TextInputFieldComponent from '../../component/Input/TextInputField';
const LoginScreen = () => {
  const {
    email,
    password,
    confirmPassword,
    showPassword,
    showConfirmPassword,
    isSignUp,
    loading,
    setEmail,
    setPassword,
    setConfirmPassword,
    toggleShowPassword,
    toggleShowConfirmPassword,
    handleSignUpToggle,
    handleLoginToggle,
    handleSignUp,
    handleLogin,
  } = useLoginScreen();

  return (
    <SafeAreaView style={styles.container}>
      <Spinner
        visible={loading}
        textContent={'Loading...'}
        textStyle={styles.spinnerTextStyle}
      />

      <Text style={styles.welcomeTitle}>{Strings.welcome}</Text>
      <Text style={styles.welcomeSubtitle}>
        {isSignUp ? 'Create your account' : 'Welcome to your Portal'}
      </Text>

      {/* Email Input */}

      <TextInputFieldComponent
        label="Email"
        value={email}
        onChangeText={setEmail}
        icon={require('../../assets/images/mail.png')}
        placeholder="your.email@gmail.com"
        enableUnderLine={false}
      />
      {/* Password Input */}
      <TextInputFieldComponent
        label="Password"
        value={password}
        onChangeText={setPassword}
        secureTextEntry={!showPassword}
        showToggle={true}
        toggleShow={toggleShowPassword}
        icon={require('../../assets/images/lock.png')}
        showPasswordIcon={require('../../assets/images/eye.png')}
        hidePasswordIcon={require('../../assets/images/eye.png')}
        placeholder="Password"
        enableUnderLine={false}
      />
      {/* Confirm Password Input - Only during Sign Up */}
      {isSignUp && (
        <TextInputFieldComponent
          label="Confirm Password"
          value={confirmPassword}
          onChangeText={setConfirmPassword}
          secureTextEntry={!showConfirmPassword}
          showToggle={true}
          toggleShow={toggleShowConfirmPassword}
          icon={require('../../assets/images/lock.png')}
          showPasswordIcon={require('../../assets/images/eye.png')}
          hidePasswordIcon={require('../../assets/images/eye.png')}
          placeholder="Confirm Password"
          enableUnderLine={false}
        />
      )}

      {/* Login/Sign Up Button */}
      <ButtonComponent
        title="Login"
        onPress={isSignUp ? handleSignUp : handleLogin}
        icon={require('../../assets/images/arrow.png')}
        enabled={true}
      />

      {/* Switch between Sign Up and Login */}
      <ButtonComponent
        title={isSignUp ? 'Back to Login' : 'Sign Up'}
        onPress={isSignUp ? handleLoginToggle : handleSignUpToggle}
        icon={require('../../assets/images/arrow.png')}
        enabled={true}
      />
    </SafeAreaView>
  );
};

export default LoginScreen;
