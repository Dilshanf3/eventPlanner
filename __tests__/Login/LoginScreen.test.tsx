import React from 'react';
import {render} from '@testing-library/react-native';
import LoginScreen from '../../src/screens/Auth/LoginScreen/LoginScreen';

// Mock the useLoginScreen hook to provide necessary props for the component
jest.mock('../../src/screens/Auth/LoginScreen/useLoginScreen', () => {
  return jest.fn(() => ({
    email: '',
    password: '',
    confirmPassword: '',
    showPassword: false,
    showConfirmPassword: false,
    isSignUp: false,
    loading: false,
    errorMessage: '',
    setEmail: jest.fn(),
    setPassword: jest.fn(),
    setConfirmPassword: jest.fn(),
    toggleShowPassword: jest.fn(),
    toggleShowConfirmPassword: jest.fn(),
    handleSignUpToggle: jest.fn(),
    handleLoginToggle: jest.fn(),
    handleSignUp: jest.fn(),
    handleLogin: jest.fn(),
  }));
});

describe('<LoginScreen />', () => {
  it('renders correctly', () => {
    const {toJSON} = render(<LoginScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
