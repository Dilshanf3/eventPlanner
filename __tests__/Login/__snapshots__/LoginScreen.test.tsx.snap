// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <Text
    style={
      {
        "color": "#191C1E",
        "fontSize": 32,
        "fontWeight": "bold",
        "marginBottom": 5,
        "textAlign": "center",
      }
    }
  >
    Welcome
  </Text>
  <Text
    style={
      {
        "color": "#757779",
        "fontSize": 16,
        "marginBottom": 40,
        "textAlign": "center",
      }
    }
  >
    Welcome to your Portal
  </Text>
  <View
    style={
      {
        "marginBottom": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#444749",
          "fontSize": 16,
          "marginBottom": 5,
        }
      }
    >
      Email
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#DA5E4214",
          "borderRadius": 5,
          "flexDirection": "row",
          "paddingHorizontal": 5,
          "textAlign": "left",
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/images/mail.png",
          }
        }
        style={
          {
            "marginHorizontal": 5,
          }
        }
      />
      <TextInput
        editable={true}
        enableUnderLine={false}
        onChangeText={[MockFunction]}
        placeholder="your.email@gmail.com"
        placeholderTextColor="#444749"
        secureTextEntry={false}
        style={
          [
            {
              "color": "#444749",
              "flex": 1,
              "fontSize": 16,
              "paddingHorizontal": 2,
              "paddingVertical": 10,
            },
            {
              "paddingLeft": 20,
            },
          ]
        }
        value=""
      />
    </View>
    <View
      style={
        {
          "backgroundColor": "#444749",
          "height": 1,
          "marginTop": -5,
        }
      }
    />
  </View>
  <View
    style={
      {
        "marginBottom": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#444749",
          "fontSize": 16,
          "marginBottom": 5,
        }
      }
    >
      Password
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#DA5E4214",
          "borderRadius": 5,
          "flexDirection": "row",
          "paddingHorizontal": 5,
          "textAlign": "left",
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../src/assets/images/lock.png",
          }
        }
        style={
          {
            "marginHorizontal": 5,
          }
        }
      />
      <TextInput
        editable={true}
        enableUnderLine={false}
        onChangeText={[MockFunction]}
        placeholder="Password"
        placeholderTextColor="#444749"
        secureTextEntry={true}
        style={
          [
            {
              "color": "#444749",
              "flex": 1,
              "fontSize": 16,
              "paddingHorizontal": 2,
              "paddingVertical": 10,
            },
            {
              "paddingLeft": 20,
            },
          ]
        }
        value=""
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "padding": 5,
            "position": "absolute",
            "right": 10,
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/images/eye.png",
            }
          }
          style={
            {
              "marginHorizontal": 5,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#444749",
          "height": 1,
          "marginTop": -5,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#DA5E42",
        "borderRadius": 5,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 20,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontSize": 16,
            "marginRight": 10,
          },
          undefined,
        ]
      }
    >
      Login
    </Text>
    <Image
      source={
        {
          "testUri": "../../../src/assets/images/arrow.png",
        }
      }
      style={
        {
          "height": 24,
          "resizeMode": "contain",
          "width": 24,
        }
      }
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#DA5E42",
        "borderRadius": 5,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": 20,
        "opacity": 1,
        "paddingVertical": 15,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontSize": 16,
            "marginRight": 10,
          },
          undefined,
        ]
      }
    >
      Sign Up
    </Text>
    <Image
      source={
        {
          "testUri": "../../../src/assets/images/arrow.png",
        }
      }
      style={
        {
          "height": 24,
          "resizeMode": "contain",
          "width": 24,
        }
      }
    />
  </View>
</RNCSafeAreaView>
`;
